### PR TITLE
Product Collection: Add new flow for adding Product Collection block

### LINKED ISSUE #42164: Product Collection - Collections - On Sale
### Background

Based on https://github.com/woocommerce/woocommerce-blocks/pull/10952 and https://github.com/woocommerce/woocommerce/issues/42166 implement a new "On Sale" collection.

### Requirements

- For now, the collection is only discoverable from the Product Collection block
- It has On Sale filter enabled by default
- On Sale filter is hidden from Inspector Controls
- It includes keyword for current block "On Sale Products"

### LINKED ISSUE #42165: Product Collection - Collections - Top Rated
### Background

Based on https://github.com/woocommerce/woocommerce-blocks/pull/10952 and https://github.com/woocommerce/woocommerce/issues/42166 implement a new "Top Rated" collection.

### Requirements

- For now, the collection is only discoverable from the Product Collection block
- It has Order By predefined to `'rating/desc'`
- Order By option should be hidden from Inspector Controls
- It includes keyword for current block "Top Rated Products"

### LINKED ISSUE #42166: Product Collection - Collections - Top Sellers
### Background

Based on https://github.com/woocommerce/woocommerce-blocks/pull/10952 implement a new "Top Sellers" collection.

### Requirements

- For now, the collection is only discoverable from the Product Collection block
- It has Order By predefined to `'popularity/desc'`
- Order By option should be hidden from Inspector Controls
- It includes keyword for current block "Best Selling Products"
- This issue includes the research regarding hiding filters from Inspector Controls depending on the active Collection - as an outcome, this logic should not only be used by this filter but also be reusable for others.

### LINKED ISSUE #42224: Product Collection: Improve the flow for adding the block
Related GitHub discussion: https://github.com/woocommerce/woocommerce-blocks/discussions/10694

### Summary

Improve the user experience for adding a new Product Collection block by introducing a placeholder state that allows the selection or creation of collections, instead of defaulting to a 3-column grid. Further, refine the overlay that appears on selecting "Choose a collection" to include patterns and presets for easy customization.  

---

### Background

The current implementation of the Product Collection block defaults to a 3-column grid with alphabetical ordering and random stock status filters ("In stock, Out of stock, On backorder"). A user can make changes via the block toolbar and the sidebar, but the initial setup might not be intuitive. We aim to bring it closer to the WordPress Core blocks' experience, which often uses placeholder states for initial configuration.  

---

### Designs

**UPDATE (13.11.23):**
 After discussing the Pattern chooser content with @manospsyx, @imanish003 and @xristos3490 we're changing the outcome of this issue. Changes:
- Skip Pattern chooser for now and stick to Collections only.
- Each Collection can have different initial layout setup (`perPage`, `pages`, `columns`).

**UPDATE (20.10.23):**
After discussing the Pattern chooser content with @manospsyx and @xristos3490 we're changing the outcome of this issue. Changes:
- Separate Collections and Patterns as separate entities
- Collections don't need previews, just icon, title and short description
- Collections don't influence the layout and don't include any additional blocks except for Product Collection (no headings, subtitles or any other blocks).
- At the moment we can stick to Collection chooser (top part of the screenshot) and pattern chooser (bottom half of screenshot) is nice to have (can be extracted to separate issue).
image

Figma: zx83mG5ZnDiacPdH6GlO2j-fi-161_23181  

- **Current default state of the Product Collection block**
image_1693827424749_0

- **Proposed placeholder state**
image_1693826998584_0

- **Proposed collection selection overlay**
image_1693827006923_0

- **Example for creating a new collection**

https://github.com/woocommerce/woocommerce-blocks/assets/191598/8fd9b551-0dfb-4f63-aef4-a12e73bd6270

---

### Acceptance Criteria

**Placeholder State**

- Implement a placeholder state. The placeholder should offer options to either choose a pre-existing collection or create a new one.

**Collection Selection Overlay**

- Upon selecting "Choose a collection", an overlay should appear.

- The overlay must contain a list of patterns and presets, including but not limited to New Arrivals, Featured, On Sale.

**Create New Collection**

- Though not finalized, explore the potential for users to define a new collection with filters and settings.

- Allow the collections to be reusable.

**Maintain Existing Functionalities**

- Ensure that the new changes do not disrupt existing functionalities like column adjustments, ordering, and filters.

---

### Feedback Loop

After the initial implementation, the @shaunandrews will review the changes. Potential adjustments to the design or functionality may be carried out based on the feedback received.  

---

### Additional Context

- There's an open question about the intersection of this new feature with WP Variations and Synced Patterns. Further exploration is needed.

CC: @kmanijak 
```[tasklist]
### Tasks
- [x] Implement the "Choose pattern" modal (part of https://github.com/woocommerce/woocommerce-blocks/issues/9372)
- [x] New custom collection adds a default Product Collection block
- [x] Add pattern: Five column grid with filters
- [x] Add pattern: Columns with filter sidebar
- [x] Add pattern: New Arrivals
- [x] Add pattern: Best Sellers
- [x] Add pattern: On sale
- [x] Implement the placeholder state
```

### PR DESCRIPTION
**Pull request by kmanijak**
## What
In this pull request, we are introducing several important changes to the WooCommerce Product Collection block. The key modifications include:
- Introducing 6 Collections:
    - New Arrivals
    - Top Rated
    - Best Selling
    - On Sale
    - Featured
    - Default Query
- Add new flow of adding Product Collection:
    - User chooses if:
        - Insert Collection - user is presented with Modal to choose Collection
        - Insert Default Product Collection - "default" Product Collection is introduced
- Adjusting existing E2E tests to the new flow and adding additional E2E tests

#### **Placeholder State**
A **Placeholder State** has been implemented for the Product Collection block. This new state provides users with a more user-friendly starting point when adding the block to a page. It offers options to either select an existing collection or add default one, streamlining the initial configuration process.
image

#### **Collection Selection Overlay**
Upon selecting "Choose Collection," an **Overlay** now appears. This overlay simplifies the process of selecting a collection by presenting users with a list of collections, including options like New Arrivals.
image

Each Collection has dedicated title, icon and description as per designs (ignore the bottom part of it):
image

#### **Collections README:**
We've added a README file for the collections of the Product Collection block. This README documents how to use collections, what attributes they support, and how to register new collections.

#### **Collections**
We've included a set of initial collections:
    - New Arrivals
    - Top Rated
    - Best Selling
    - On Sale
    - Featured
    - Default Query

Fixes woocommerce/woocommerce#42224
Fixes https://github.com/woocommerce/woocommerce/issues/42166
Fixes https://github.com/woocommerce/woocommerce/issues/42165
Fixes https://github.com/woocommerce/woocommerce/issues/42164

## Why

The primary motivation behind these changes is to significantly improve the user experience when adding a new Product Collection block. The existing implementation defaults to a 3-column grid with alphabetical ordering and random stock status filters, which might not be intuitive for users. We are introducing the following enhancements:

These changes are designed to make the Product Collection block more accessible and user-centric.

For more info, you may also check Github Discussion #10694. 

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

### Case 1: "Add default Product Collection" flow
- Create a new post & add `Product Collection (Beta)` block to it.
- Check that the block displays a placeholder state as shown in

image

- Click the "Use default collection" button.
- Verify that no errors occur and the block renders correctly in the editor.
- Product Collection block is added with:
  - 9 products per page,
  - 3 columns
  - unlimited pages (pages set to 0)
  - there are "Pagination" and "No Results" blocks included
  - no heading
  - no custom filters enabled
- Publish the post & verify that block looks similar to Editor on Frontend.
- Make random changes to the block's filters and settings in the editor, save them, and ensure the changes are reflected correctly on the frontend.
- Repeat the steps in Product Catalog template - after injecting, Product Collection should have "Inherit query from template" option enabled by default

### Case 2: "Choose Collection" flow 
- Create a new post & add `Product Collection (Beta)` block to it.
- In the placeholder state, click the "Choose Collection" button.
- Confirm that a modal window appears with 6 Collections as on a screenshot:
image

- Choose "New Arrivals" and click "Continue"
- Verify that the selected collection displays correctly in the editor.
- Expected settings
  - The same for each Collection:
    - 5 columns
    - 5 products per page
    - Single page
    - No "Pagination" nor "No Results" blocks
  - Per Collection:

Collection | Order | Other settings | Heading
-- | -- | -- | --
New Arrivals | Newest to oldest | - | Newest products
Top Rated | Top Rated | - | Top rated products
Best Selling | Best Selling | - | Best selling products
On Sale | A -> Z | "Show only products on sale" enabled | On sale products
Featured | A -> Z | "Show only featured products" enabled | Featured products

- Publish the post & Confirm that the selected pattern renders correctly on the frontend.

### Case 3: "Choose Collection" flow with "All Products"
- Create a new post & add `Product Collection (Beta)` block to it.
- In the placeholder state, click the "Choose Collection" button.
- Choose "All Products" and click "Continue"
- Verify that no errors occur and the block renders correctly in the editor.
- Product Collection block is added with 9 products per page, no heading and no custom filters enabled
- Publish the post & verify that block looks similar to Editor on Frontend.
- Make random changes to the block's filters and settings in the editor, save them, and ensure the changes are reflected correctly on the frontend.
- Repeat the steps in Product Catalog template - after injecting, Product Collection should have "Inherit query from template" option enabled by default

### Case 4: Verify Collections ARE NOT available in the inserter
- In any place in editor, try to add the blocks:
  - New Arrivals
  - Top Rated
  - Best Selling
  - On Sale
  - Featured
- Expected: they ARE NOT discoverable in the inserter.

### Misc
- Feel free to leave a comment if you encounter any issues or notice any other problems while testing this PR. Although we have thoroughly tested this PR, it's possible that we might have missed something.

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Product Collection: Added new flow to choose patterns/collections while adding the block

### CODE REVIEW

**Comment by shaunandrews** [comment]
CleanShot 2023-09-21 at 09 23 19@2x

The design has the secondary action in the placeholder labeled "New custom collection" — but I understand we don't have that feature yet. The "Add default Product Collection" label is a bit wordy. Maybe we can change it to something shorter, like: "Use default collection"

--

CleanShot 2023-09-21 at 09 27 14@2x

The "Choose a pattern" modal is yuuuuge! It's nice to have large previews, but they tend to flow together — it's really difficult to identify each individual option. It's also a little confusing to mix up the terms "collection" and "pattern."

Is there any way we can get this closer to design below?

CleanShot 2023-09-21 at 09 28 56@2x

**Comment by imanish003** [comment]
Hi @shaunandrews, Thanks for the quick feedback 🙏 

> The design has the secondary action in the placeholder labeled "New custom collection" — but I understand we don't have that feature yet. The "Add default Product Collection" label is a bit wordy. Maybe we can change it to something shorter, like: "Use default collection"  

Agreed! I will change it to "Use default collection."

> It's also a little confusing to mix up the terms "collection" and "pattern."  

Agreed! I thought about changing the title of the modal from "Choose a pattern" to "Choose a collection," but since it contains both, I wasn't sure what the correct title should be. However, I will go ahead and match it with the design. Thanks for pointing this out.

> Is there any way we can get this closer to design below?  

Let me investigate it today. I think this is doable.

**Comment by imanish003** [comment]
We're utilizing `BlockPatternList` from `@wordpress/components`, which currently restricts us from achieving the desired UI for our block.

@karolmanijak, I'm contemplating whether we should contribute to the upstream or craft a custom component specifically for our patterns list.
- Regarding the limitations and possibilities with `BlockPatternList`, here's what I've found:
	- One available customization is the `viewportWidth` property for each pattern, allowing us to control its width. For example:
	    
	  ```jsx
	  const allPatterns = [
	    { name: 'pattern-1', viewportWidth: 500, /* other properties */ },
	    { name: 'pattern-2', viewportWidth: 800, /* other properties */ },
	    // ... additional patterns
	  ];
	  <BlockPatternList
	    shownPatterns={allPatterns}
	    // ... other props
	  />
	  
	  ```
	- Beyond this, there doesn't seem to be any additional properties that facilitate the styling of individual Pattern Previews.
- Should we decide to make contributions upstream, the following changes would be necessary:
	- Upon reviewing `BlockPatternList`'s internal implementation in Gutenberg, it's clear that it employs the `BlockPreview` component. The `BlockPreview` does have an `additionalStyles` prop that can be utilized for custom styling.
	- A possible approach could be contributing an update that allows `BlockPatternList` to accept an `additionalStyles` prop for each `shownPatterns`. This would then be passed to `BlockPreview`, enabling us to tailor the appearance of each Pattern within the list.

I welcome your insights on this matter. Do you foresee any issues with the suggested approach of incorporating the `additionalStyles` prop?

### Resources
- [BlockPreview Documentation](https://github.com/WordPress/gutenberg/blob/3e497b3df6c7d18f99fa572eb4cec7237475ee1a/packages/block-editor/src/components/block-preview/README.md)
- [BlockPatternList Documentation](https://github.com/WordPress/gutenberg/blob/3e497b3df6c7d18f99fa572eb4cec7237475ee1a/packages/block-editor/src/components/block-patterns-list/README.md)

**Comment by kmanijak** [comment]
@imanish003, alternatively we could extend the `/assets/js/blocks/product-collection/edit/editor.scss` file with custom CSS, like:

```
.wc-blocks-product-collection__selection-modal {
         ...
   	.block-editor-block-preview__container {
		height: 200px;
		align-items: flex-start;
	}
}
```

Which results in such a view:

image

Let's keep in mind we already have [a custom code](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/product-collection/editor.scss#L14) responsible for splitting the collections into columns, and it's quite common we apply custom styles to `@wordpress/components`. Hence I think it's a low-effort solution that gives us the expected result. What do you think?

**Comment by imanish003** [comment]
Hey Karol 👋 

I like the idea. Although I try to avoid using CSS as much as possible, it makes sense to use CSS in this case. Additionally, I never intuitively think about CSS solutions 🤦‍♂️.

**Comment by kmanijak** [comment]
@shaunandrews, here's how the modal looks right now:

Desktop | Mobile
-- | --
image | image

Differences from designs:
1. I didn't apply the padding around the Collection previews because the preview is scaled down dynamically depending on the container size, without taking padding into account. That means it requires hacky solutions and can easily break if there's any change in Gutenberg. I applied padding on top to give it more space though.
2. No "Cancel" and "Continue" buttons. I followed here the Query Loop behavior, where clicking a pattern applies it instantly. Let me know if you think we should actually go with the buttons.

**Comment by imanish003** [comment]
@kmanijak  I've tested the functionality, and everything is working as expected. Excellent work! 🚀

- We might consider setting the "Shrink columns to fit" option to true for all collections.
    - Alternatively, as I previously mentioned in slack thread(p1700116561625059/1700055115.096619-slack-C05UMTA7W0P), it could be a good idea to set this as the default whenever a Product Collection block is added.

I'll start reviewing the code now, but I might not finish today as it's almost the end of my workday 🙂.

**Review by imanish003 — COMMENTED** [review]
@kmanijak  I've tested the functionality, and everything is working as expected. Excellent work! 🚀

- We might consider setting the "Shrink columns to fit" option to true for all collections.
    - Alternatively, as I previously mentioned in slack thread(p1700116561625059/1700055115.096619-slack-C05UMTA7W0P), it could be a good idea to set this as the default whenever a Product Collection block is added.

I'll start reviewing the code now, but I might not finish today as it's almost the end of my workday 🙂.

**Comment by kmanijak** [comment]
> We might consider setting the "Shrink columns to fit" option to true for all collections

@imanish003  I started doing that but I figured it would be better to extract to separate PR: https://github.com/woocommerce/woocommerce-blocks/pull/11821.

This PR is already huge and even though enabling "Shrink columns to fit" is small in code, it would add couple of more testing scenarios. If https://github.com/woocommerce/woocommerce-blocks/pull/11821 gets merged before this PR, I'll update it to enable the option in Collections. Otherwise, I'll update https://github.com/woocommerce/woocommerce-blocks/pull/11821 to include Collections.

**Comment by imanish003** [comment]
@kmanijak That makes sense 🤝

**Comment by kmanijak** [comment]
This work awaits a designer's review/input. For the time being it's paused.

**Comment by kmanijak** [comment]
Bug: All Products have "inherit query from template" hidden the same as existing collections.

Expected changes:

- "Choose Collection" needs to be "Choose collection"
- "Use default collection" can change to "Create custom"
- Choose from pre-existing collections or add default one." becomes "Choose a product collection to display, or create your own."

EDIT: Done ✅

**Comment by kmanijak** [comment]
Closing. It's been moved to monorepo: https://github.com/woocommerce/woocommerce/pull/42696
